File: samples/python/bidi-sample/bidi_sample.py
import asyncio
import sys

import websockets

from aeron_cache.bidi_client import AeronBidiClient, BidiError
from aeron_cache.models import (
    BulkCacheOpsRequest,
    BulkOperationType,
    CacheOperationRequest,
)


async def main():
    ws_url = "ws://localhost:7071"
    if len(sys.argv) > 1:
        ws_url = sys.argv[1]

    print(f"Starting Bidi (bidirectional WebSocket) Sample against ws={ws_url}")

    cache_id = "bidi-sample-cache"
    counter_cache_id = "bidi-sample-counter-cache"

    try:
        async with AeronBidiClient(ws_url) as client:
            # --- Cache lifecycle ---
            create = await client.create_cache(cache_id)
            print(f"Created cache: {create.cacheId} (status={create.operationStatus})")

            # --- put + get ---
            print("Putting key 'greeting' -> 'hello'")
            put = await client.put_item(cache_id, "greeting", "hello")
            print(f"Put operation status: {put.operationStatus}")

            got = await client.get_item(cache_id, "greeting")
            print(f"Read key 'greeting': {got.value or 'not found'}")

            # --- patch (merge) ---
            print("Putting key 'doc' -> {\"a\":1}")
            await client.put_item(cache_id, "doc", '{"a":1}')
            print("Patching key 'doc' with {\"b\":2}")
            await client.patch_item(cache_id, "doc", '{"b":2}')
            merged = await client.get_item(cache_id, "doc")
            print(f"Merged 'doc' value: {merged.value}")

            # --- counter demo ---
            counter_create = await client.create_counter_cache(counter_cache_id)
            print(f"Created counter cache: {counter_create.cacheId} (status={counter_create.operationStatus})")

            print("Putting counter 'requests' -> 10")
            await client.put_counter(counter_cache_id, "requests", 10)
            inc = await client.increment_counter(counter_cache_id, "requests", 5)
            print(f"Incremented 'requests' by 5 -> {inc.value}")
            counter = await client.get_counter(counter_cache_id, "requests")
            print(f"Read counter 'requests': {counter.value}")

            # --- bulk read + stats ---
            items = await client.get_cache_items(cache_id)
            print(f"Cache items ({len(items.items)}):")
            for item in items.items:
                print(f"  {item.key} = {item.value}")

            stats = await client.get_stats()
            print(f"Cache stats ({len(stats)} entries):")
            for stat in stats:
                print(f"  {stat.cacheId}: size={stat.size} added={stat.addedCount} removed={stat.removedCount}")

            # --- bulk operations ---
            # A single bulk frame carries a batch of operations; the server streams back
            # per-operation results, each echoing its own requestId.
            bulk_request = BulkCacheOpsRequest(
                requestId="bidi-bulk-1",
                operations=[
                    CacheOperationRequest(
                        operationType=BulkOperationType.ADD_ITEM,
                        requestId="op-1", cacheId=cache_id, key="bk1", value="bv1"
                    ),
                    CacheOperationRequest(
                        operationType=BulkOperationType.ADD_ITEM,
                        requestId="op-2", cacheId=cache_id, key="bk2", value="bv2"
                    ),
                    CacheOperationRequest(
                        operationType=BulkOperationType.GET_ITEM,
                        requestId="op-3", cacheId=cache_id, key="bk1"
                    ),
                ],
            )
            bulk_response = await client.bulk_ops(bulk_request)
            print("Bulk operation responses:")
            for op_resp in bulk_response.operationResponses:
                suffix = f" ({op_resp.value})" if op_resp.value else ""
                print(f"  {op_resp.requestId} -> {op_resp.status}{suffix}")

            # --- timers ---
            # A timed entry schedules a pending TTL removal timer; get_timers lists all pending
            # timers (cache + counter), each tagged with its type.
            await client.put_timed_item(cache_id, "expiring", "gone-soon", 600000)
            print("Listing all pending TTL timers:")
            for timer in await client.get_timers():
                print(f"  [{timer.timerType}] {timer.cacheId}/{timer.key} fires at {timer.deadline}")

            # --- live subscription ---
            received = asyncio.Event()
            last_event = {}

            def on_event(event):
                last_event["event"] = event
                received.set()

            print("Subscribing to live updates on 'bidi-sample-cache'")
            sub = await client.subscribe(cache_id, on_event)

            print("Putting key 'live' -> 'streamed' to trigger an event")
            await client.put_item(cache_id, "live", "streamed")

            try:
                await asyncio.wait_for(received.wait(), timeout=5.0)
                event = last_event.get("event")
                print(
                    f"Received live event: type={event.eventType} "
                    f"key={event.itemKey} value={event.itemValue}"
                )
            except asyncio.TimeoutError:
                print("No live event received within 5s")

            await sub.close()
            print("Closed subscription")

            # --- teardown ---
            deleted = await client.delete_cache(cache_id)
            print(f"Deleted cache: {deleted.cacheId} (status={deleted.operationStatus})")
            deleted_counters = await client.delete_counter_cache(counter_cache_id)
            print(f"Deleted counter cache: {deleted_counters.cacheId} (status={deleted_counters.operationStatus})")

    except (OSError, ConnectionError, websockets.exceptions.WebSocketException) as e:
        print(f"Could not connect to the bidi WebSocket at {ws_url}: {e}")
        print("Is the backend running (HTTP 7070, WS 7071)?")
        sys.exit(1)
    except BidiError as e:
        print(f"Server returned an error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: samples/java/bidi-sample/src/main/java/com/aeron/cache/sample/BidiSample.java
package com.aeron.cache.sample;

import com.bhf.aeroncache.client.bidi.AeronBidiClient;
import com.bhf.aeroncache.client.bidi.BidiSubscription;
import com.bhf.aeroncache.models.BulkCacheOpsRequest;
import com.bhf.aeroncache.models.BulkOperationType;
import com.bhf.aeroncache.models.CacheItem;
import com.bhf.aeroncache.models.CacheOperationRequest;
import com.bhf.aeroncache.models.CacheOperationResponse;
import com.bhf.aeroncache.models.StatEntry;
import com.bhf.aeroncache.models.TimerInfo;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * Demonstrates the bidirectional WebSocket transport — the JSON/WebSocket analogue of the Aeron
 * gateway client, carried over a single persistent WebSocket connection ({@code /api/ws/v1/bidi}).
 * <p>
 * Point it at the backend's bidi WebSocket endpoint with {@code args[0]} (default
 * {@code ws://localhost:7071}). The backend must be running with the WebSocket transport enabled.
 */
public class BidiSample {

    public static void main(String[] args) throws Exception {
        final String wsUrl = args.length > 0 ? args[0] : "ws://localhost:7071";
        System.out.println("Starting Bidi Sample against " + wsUrl);

        try (AeronBidiClient client = new AeronBidiClient(wsUrl)) {
            client.connect();
            System.out.println("Connected to bidi WebSocket endpoint.");

            // --- Cache operations ---
            final String cacheId = "bidi-sample-cache";
            System.out.println();
            System.out.println("--- Cache ---");
            System.out.println("Created cache: " + client.createCache(cacheId).getCacheId());
            System.out.println("Put bidi-key -> bidi-value: "
                    + client.putItem(cacheId, "bidi-key", "bidi-value").getOperationStatus());
            System.out.println("Get bidi-key -> " + client.getItem(cacheId, "bidi-key").getValue());

            // --- Patch (JSON merge) ---
            // Put a base document, then patch it: the server merges the patch into the stored value.
            System.out.println();
            System.out.println("--- Patch ---");
            client.putItem(cacheId, "doc", "{\"a\":1}");
            System.out.println("Put doc -> {\"a\":1}");
            System.out.println("Patch doc <- {\"b\":2}: "
                    + client.patchItem(cacheId, "doc", "{\"b\":2}").getOperationStatus());
            System.out.println("Merged doc -> " + client.getItem(cacheId, "doc").getValue());

            // --- Counters ---
            System.out.println();
            System.out.println("--- Counters ---");
            final String counterCache = "bidi-sample-counters";
            System.out.println("Created counter cache: " + client.createCounterCache(counterCache).getCacheId());
            System.out.println("Put hits = 10: " + client.putCounter(counterCache, "hits", 10).getOperationStatus());
            System.out.println("increment hits +5 -> " + client.incrementCounter(counterCache, "hits", 5).getValue());
            System.out.println("Get hits -> " + client.getCounter(counterCache, "hits").getValue());

            // --- Bulk read + stats ---
            System.out.println();
            System.out.println("--- Entries & stats ---");
            System.out.println("Cache items:");
            for (CacheItem item : client.getCacheItems(cacheId).getItems()) {
                System.out.println("  " + item.getKey() + " = " + item.getValue());
            }
            System.out.println("Cache stats:");
            for (StatEntry stat : client.getStats()) {
                System.out.println("  " + stat.getCacheId() + " size=" + stat.getSize()
                        + " added=" + stat.getAddedCount() + " removed=" + stat.getRemovedCount());
            }

            // --- Bulk operations ---
            // A single `bulk` frame carries a batch of operations (regular-cache and counter ops may be
            // mixed); the server streams back per-operation results, each echoing its own requestId.
            System.out.println();
            System.out.println("--- Bulk operations ---");
            final BulkCacheOpsRequest bulk = BulkCacheOpsRequest.builder()
                    .requestId("bidi-bulk-1")
                    .addOperation(CacheOperationRequest.builder()
                            .operationType(BulkOperationType.ADD_ITEM)
                            .requestId("op-1").cacheId(cacheId).key("bk1").value("bv1").build())
                    .addOperation(CacheOperationRequest.builder()
                            .operationType(BulkOperationType.ADD_ITEM)
                            .requestId("op-2").cacheId(cacheId).key("bk2").value("bv2").build())
                    .addOperation(CacheOperationRequest.builder()
                            .operationType(BulkOperationType.GET_ITEM)
                            .requestId("op-3").cacheId(cacheId).key("bk1").build())
                    .build();
            for (CacheOperationResponse op : client.bulkOps(bulk).getOperationResponses()) {
                System.out.println("  " + op.getRequestId() + " -> " + op.getStatus()
                        + (op.getValue() != null ? " (" + op.getValue() + ")" : ""));
            }

            // --- Timers ---
            // A timed entry schedules a pending TTL removal timer; getTimers lists all pending timers
            // (cache + counter), each tagged with its type.
            System.out.println();
            System.out.println("--- Timers ---");
            client.putTimedItem(cacheId, "expiring", "gone-soon", 600_000);
            for (TimerInfo timer : client.getTimers()) {
                System.out.println("  [" + timer.getTimerType() + "] " + timer.getCacheId()
                        + "/" + timer.getKey() + " fires at " + timer.getDeadline());
            }

            // --- Live subscription ---
            // Subscribe over the same connection, then write a key and observe the streamed update.
            System.out.println();
            System.out.println("--- Live subscription ---");
            final CountDownLatch received = new CountDownLatch(1);
            try (BidiSubscription subscription = client.subscribe(cacheId, event -> {
                System.out.println("  [update] " + event.getEventType() + " "
                        + event.getItemKey() + "=" + event.getItemValue());
                received.countDown();
            })) {
                // Let the subscription establish before writing, so we observe our own update.
                Thread.sleep(500);
                client.putItem(cacheId, "streamed-key", "streamed-value");
                if (!received.await(5, TimeUnit.SECONDS)) {
                    System.out.println("  (no streamed update received within 5s)");
                }
            }
            System.out.println("Subscription closed.");

            // --- Cleanup ---
            System.out.println();
            client.deleteCache(cacheId);
            client.deleteCounterCache(counterCache);
            System.out.println("Deleted caches. Done.");
        } catch (Exception e) {
            System.err.println("Bidi sample failed — is the backend running with the WebSocket "
                    + "transport reachable at " + wsUrl + "?");
            System.err.println("  cause: " + e.getMessage());
        }
    }
}
